sentry-core/src/main/java/io/sentry/core/Hub.java

```diff
@@ -1,13 +1,34 @@
 package io.sentry.core;
 
 import io.sentry.core.protocol.SentryId;
+import java.util.ArrayDeque;
+import java.util.Deque;
 
 public class Hub implements IHub {
+
+  private class StackItem {
+    public ISentryClient client;
+    public Scope scope;
+
+    public StackItem(ISentryClient client, Scope scope) {
+      this.client = client;
+      this.scope = scope;
+    }
+  }
+
   private SentryOptions options;
   private volatile boolean isEnabled;
+  private Deque<StackItem> stack = new ArrayDeque<>();
```

Comment:
why don't we use an eg. `ConcurrentLinkedDeque` instead of creating a `locker`?

Reply:
I'm holding a lock while I do more than one operation with it like:

```java
    synchronized (lock) {
      StackItem item = stack.peek();
      item.client.close(options.getShutdownTimeout());
    }
```

So that it synchronizes with `bindClient`.

Scope and Client must be thread-safe so the question is: Do we care changes to client/scope are atomic?

Reply:
got it, I'd not know how to answer it.

In case we need to lock the operations itself, instead of creating our own lock Object, depending on the use-case, these ones could be a bit more performatic `ReentrantLock/ReadWriteLock`

Reply:
Given that a single Hub is not supposed to be used by more than 1 thread I would not optimize for correctness of calls to Hub.something to be atomic with regards to client and scope pair but rather just make sure it is thread-safe. Meaning calls to Hub done concurrently will not throw.

That said, IMO a immutable or thread-safe collection is good enough here.

Reply:
Unfortunately ConcurrentLinkedDeque requires API 21

Reply:
@bruno-garcia LinkedBlockingDeque requires API 9, it's also thread-safe afaik

Reply:
Thanks. I'll look at it